src/app/clusters/color-control-server/color-control-server.cpp

```diff
@@ -1498,12 +1500,150 @@ bool emberAfColorControlClusterStopMoveStepCallback(EndpointId aEndpoint, app::C
+    colorHueTransitionState.currentEnhancedHue = currentHue;
+
+    if (direction)
+    {
+        colorHueTransitionState.finalEnhancedHue = static_cast<uint16_t>(startHue - 1);
+    }
+    else
+    {
+        colorHueTransitionState.finalEnhancedHue = static_cast<uint16_t>(startHue + 1);
+    }
+
+    colorHueTransitionState.up     = direction;
+    colorHueTransitionState.repeat = true;
+
+    colorHueTransitionState.stepsRemaining = static_cast<uint16_t>(time * TRANSITION_TIME_1S);
+    colorHueTransitionState.stepsTotal     = static_cast<uint16_t>(time * TRANSITION_TIME_1S);
+    colorHueTransitionState.endpoint       = endpoint;
+
+    writeRemainingTime(endpoint, MAX_INT16U_VALUE);
+    emberEventControlSetDelayMS(&COLOR_HSV_CONTROL, UPDATE_TIME_MS);
+}
+
+bool emberAfColorControlClusterColorLoopSetCallback(chip::EndpointId aEndpoint, chip::app::CommandHandler * commandObj,
+                                                    uint8_t updateFlags, uint8_t action, uint8_t direction, uint16_t time,
+                                                    uint16_t startHue, uint8_t optionsMask, uint8_t optionsOverride)
+{
+    EndpointId endpoint = emberAfCurrentEndpoint();
+    uint8_t isColorLoopActive;
+    uint8_t deactiveColorLoop;
+
+    if (!shouldExecuteIfOff(endpoint, optionsMask, optionsOverride))
+    {
+        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
+        return true;
+    }
+
+    isColorLoopActive = 0;
+    Attributes::GetColorLoopActive(endpoint, &isColorLoopActive);
+
+    deactiveColorLoop =
+        (updateFlags & EMBER_AF_COLOR_LOOP_UPDATE_FLAGS_UPDATE_ACTION) && (action == EMBER_ZCL_COLOR_LOOP_ACTION_DEACTIVATE);
+
+    if (updateFlags & EMBER_AF_COLOR_LOOP_UPDATE_FLAGS_UPDATE_DIRECTION)
+    {
+        Attributes::SetColorLoopDirection(endpoint, direction);
+
+        // Checks if color loop is active and stays active
+        if (isColorLoopActive && !deactiveColorLoop)
+        {
+            colorHueTransitionState.up                 = direction;
+            colorHueTransitionState.initialEnhancedHue = colorHueTransitionState.currentEnhancedHue;
+
+            if (direction)
+            {
+                colorHueTransitionState.finalEnhancedHue = static_cast<uint16_t>(colorHueTransitionState.initialEnhancedHue - 1);
+            }
+            else
+            {
+                colorHueTransitionState.finalEnhancedHue = static_cast<uint16_t>(colorHueTransitionState.initialEnhancedHue + 1);
+            }
+            colorHueTransitionState.stepsRemaining = colorHueTransitionState.stepsTotal;
+        }
+    }
+
+    if (updateFlags & EMBER_AF_COLOR_LOOP_UPDATE_FLAGS_UPDATE_TIME)
+    {
+        Attributes::SetColorLoopTime(endpoint, time);
+
+        // Checks if color loop is active and stays active
+        if (isColorLoopActive && !deactiveColorLoop)
+        {
+            colorHueTransitionState.stepsTotal = static_cast<uint16_t>(time * TRANSITION_TIME_1S);
```

Comment:
If stepsRemaining is larger than this new time, will all the code that touches those be OK?  Or should this update stepsRemaining as needed too? I'd think we don't want to end up in an unexpected state here....

Reply:
Best option is to do the same thing as the direction attribute. Avoids all possible edge cases since only changing the stepsRemaining will cause a faulty calculation on the next hue update.